sweeper/sweeper.go

```diff
@@ -0,0 +1,191 @@
+package sweeper
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/skynetlabs/pinner/database"
+	"github.com/skynetlabs/pinner/skyd"
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/SkynetLabs/skyd/build"
+	"gitlab.com/SkynetLabs/skyd/skymodules"
+)
+
+type (
+	// Schedule defines how often, if at all, we sweep this server automatically.
+	Schedule struct {
+		Period   time.Duration
+		cancelCh chan struct{}
+	}
+	// Status represents the status of a sweep.
+	Status struct {
+		InProgress bool
+		Error      error
+		StartTime  time.Time
+		EndTime    time.Time
+	}
+	// Sweeper takes care of sweeping the files pinned by the local skyd server
+	// and marks them as pinned by the local server.
+	Sweeper struct {
+		staticDB         *database.DB
+		staticSkydClient skyd.Client
+		staticServerName string
+
+		schedule   *Schedule
+		scheduleMu sync.Mutex
+
+		status   Status
+		statusMu sync.Mutex
+	}
+)
+
+// New returns a new Sweeper.
+func New(db *database.DB, skydc skyd.Client, serverName string) *Sweeper {
+	return &Sweeper{
+		staticDB:         db,
+		staticSkydClient: skydc,
+		staticServerName: serverName,
+	}
+}
+
+// Schedule a scan to run on each period.
+func (s *Sweeper) Schedule(period time.Duration) {
+	s.scheduleMu.Lock()
+	defer s.scheduleMu.Unlock()
+	if s.schedule != nil {
+		close(s.schedule.cancelCh)
+	}
+	s.schedule = &Schedule{
+		Period:   period,
+		cancelCh: make(chan struct{}),
+	}
+	go func() {
```

Comment:
This could be a separate `threadedSweepLoop` function.

Reply:
I'm not sure that's a good idea. On one hand, we're not reusing this anywhere and it's a very short and straightforward function. On the other hand, it interacts with the `s.schedule` struct which we set up here. I'm not sure I'd like to even make it possible for this logic to be reused.

What is the benefit which you have in mind for moving this to a named function?

Reply:
I took a second look here and we don't need a goroutine at all. The cache rebuild happens in its own goroutine anyway.

Reply:
The benefit is mostly readability and documentability (not sure that is a word lol) but it's not a dealbreaker for this PR.